### PR TITLE
Migration/beam

### PR DESCRIPTION
**Pull request by rvilarl**
Beam also there!

### CODE REVIEW

File: src/beam.ts [hunk 1]
```diff
@@ -10,8 +10,44 @@ import { Element } from './element';
...
+  max_slope: number;
+  min_slope: number;
+  slope_iterations: number;
+  slope_cost: number;
+  stemlet_extension: number;
+  partial_beam_length: number;
+  min_flat_beam_offset: number;
+}
+
+export interface GenerateBeamConfig {
```
Comment (0xfe) [inline src/beam.ts:38]: This can be inlined.

File: src/beam.ts [hunk 2]
```diff
@@ -675,19 +732,19 @@ export class Beam extends Element {
   }
 
   // Get the x coordinates for the beam lines of specific `duration`
-  getBeamLines(duration) {
+  getBeamLines(duration: string): BeamInfo[] {
```
Comment (0xfe) [inline src/beam.ts:735]: No need for separate exported type `BeamInfo` here, okay to return `{ start: number, end?: number }[]`

File: src/beam.ts [hunk 3]
```diff
@@ -675,19 +732,19 @@ export class Beam extends Element {
...
     const tick_of_duration = Flow.durationToTicks(duration);
-    const beam_lines = [];
+    const beam_lines: BeamInfo[] = [];
```
Comment (0xfe) [inline src/beam.ts:737]: As above, no need to define exported interface.

Here you can define a local type:

```typescript
type BeamInfo = { start: number, end?: number };
const beam_lines: BeamInfo[] = [];
...
```

File: src/beam.ts [hunk 4]
```diff
@@ -10,8 +10,44 @@ import { Element } from './element';
...
+import { StemmableNote } from './stemmablenote';
+import { Voice } from './voice';
+import { RenderContext } from './types/common';
+
+export interface BeamInfo {
+  start: number;
+  end?: number;
+}
+
+export interface BeamRenderOptions {
```
Comment (0xfe) [inline src/beam.ts:23]: This can be inline too. No need to export.

File: src/beam.ts (hunk 1 above)
Comment (rvilarl) [reply src/beam.ts:38]: Do you mean to inline GenerateBeamConfig? The structure has 9 attributes.

File: src/beam.ts (hunk 2 above)
Comment (rvilarl) [reply src/beam.ts:735]: If I define the type, couldn´t I use it here as well?

File: src/beam.ts (hunk 3 above)
Comment (rvilarl) [reply src/beam.ts:737]: Global not exported is also fine?

File: src/beam.ts (hunk 4 above)
Comment (rvilarl) [reply src/beam.ts:23]: This one is 11 attributes.

File: src/beam.ts (hunk 2 above)
Comment (rvilarl) [reply src/beam.ts]: ok you wanted the type local

File: src/beam.ts (hunk 3 above)
Comment (rvilarl) [reply src/beam.ts:733]: done local

File: src/beam.ts (hunk 4 above)
Comment (rvilarl) [reply src/beam.ts:18]: Inline?

**Comment by rvilarl** [comment]
@0xfe on the both inline before I wait for your confirmation. I think that both interfaces have too many attributes to inline them.
The local BeamInfo chage is done.

File: src/beam.ts (hunk 1 above)
Comment (0xfe) [reply src/beam.ts:33]: Yes, correct -- the structure as a whole is only used once, so I prefer to keep it near where it's used.

File: src/beam.ts (hunk 4 above)
Comment (0xfe) [reply src/beam.ts:18]: Same -- yes, inline please.

File: src/beam.ts [hunk 5]
```diff
@@ -635,25 +684,28 @@ export class Beam extends Element {
...
-
-      note.stem.setExtension(preBeamExtension + beamExtension);
-      note.stem.renderHeightAdjustment = -Stem.WIDTH / 2;
-
-      if (note.isRest() && show_stemlets) {
-        const beamWidth = beam_width;
-        const totalBeamWidth = (beam_count - 1) * beamWidth * 1.5 + beamWidth;
-        note.stem.setVisibility(true).setStemlet(true, totalBeamWidth + stemlet_extension);
+      const stem = note.getStem();
+      if (stem) {
```
Comment (0xfe) [inline src/beam.ts:688]: Add an `else` here to throw an exception, so it doesn't silently fail.

File: src/beam.ts [hunk 6]
```diff
@@ -675,19 +727,20 @@ export class Beam extends Element {
   }
 
   // Get the x coordinates for the beam lines of specific `duration`
-  getBeamLines(duration) {
+  getBeamLines(duration: string): { start: number; end?: number }[] {
+    type BeamInfo = { start: number; end?: number };
```
Comment (0xfe) [inline src/beam.ts:731]: Can you move this closer to the usage:

```typescript
getBeamLines(...) {
  const tick_of_duration ...
  let beam_started ...

  type BeamInfo ...
  const beam_lines: BeamInfo[]...
  let current_beam: BeamInfo...
```

File: src/beam.ts [hunk 7]
```diff
@@ -816,30 +868,32 @@ export class Beam extends Element {
...
-        const lastBeamY = this.getSlopeY(lastBeamX, firstStemX, beamY, this.slope);
-
-        this.context.beginPath();
-        this.context.moveTo(startBeamX, startBeamY);
-        this.context.lineTo(startBeamX, startBeamY + beamThickness);
-        this.context.lineTo(lastBeamX + 1, lastBeamY + beamThickness);
-        this.context.lineTo(lastBeamX + 1, lastBeamY);
-        this.context.closePath();
-        this.context.fill();
+        if (lastBeamX) {
```
Comment (0xfe) [inline src/beam.ts:871]: Exception if not defined.

Comment (rvilarl) [reply src/beam.ts:871]: ok

File: src/beam.ts (hunk 6 above)
Comment (rvilarl) [reply src/beam.ts:731]: done

File: src/beam.ts (hunk 5 above)
Comment (rvilarl) [reply src/beam.ts:688]: ok

**Comment by rvilarl** [comment]
Done

File: src/beam.ts [hunk 8]
```diff
@@ -39,16 +43,45 @@ const BEAM_LEFT = 'L';
 const BEAM_RIGHT = 'R';
 const BEAM_BOTH = 'B';
 
+type tDefaults = {
```
Comment (0xfe) [inline src/beam.ts:46]: This can be inlined.

**Review by 0xfe — COMMENTED** [review]
Looks good. Just one minor nit.

File: src/beam.ts (hunk 8 above)
Comment (rvilarl) [reply src/beam.ts]: done

**Comment by rvilarl** [comment]
Ready to merge

**Comment by 0xfe** [comment]
Great!

**Comment by ronyeh** [comment]
@rvilarl  are you doing a big migration to TypeScript with full type annotations? Exciting!

@0xfe @rvilarl Let me know if I can help with testing or anything else. I'm comfortable with TypeScript. :-)

**Comment by mscuthbert** [comment]
> @rvilarl are you doing a big migration to TypeScript with full type annotations? Exciting!
> 
> @0xfe @rvilarl Let me know if I can help with testing or anything else. I'm comfortable with TypeScript. :-)

Indeed he is, but help is always welcome.  He's doing one module at a time, and it's going great!

**Comment by 0xfe** [comment]
@ronyeh Thanks. Most helpful right now would be testing your applications against HEAD as the migration occurs, and reporting if you observe anything unexpected.

**Comment by rvilarl** [comment]
> @ronyeh Thanks. Most helpful right now would be testing your applications against HEAD as the migration occurs, and reporting if you observe anything unexpected.

@ronyeh  no doubt this also helps a lot :)

**Pull request merged**